src/recipe_dictionary.cpp

```diff
@@ -1,95 +1,289 @@
+            auto arr = bp.next_array();
+            r.byproducts[ arr.get_string( 0 ) ] += arr.size() == 2 ? arr.get_int( 1 ) : 1;
+        }
+    }
+
+    if( jo.has_member( "book_learn" ) ) {
+        auto bk = jo.get_array( "book_learn" );
+        r.booksets.clear();
+
+        while( bk.has_more() ) {
+            auto arr = bk.next_array();
+            r.booksets.emplace( arr.get_string( 0 ), arr.get_int( 1 ) );
+        }
+    }
+
+    if( jo.has_string( "using" ) ) {
+        r.reqs = { { requirement_id( jo.get_string( "using" ) ), 1 } };
+
+    } else if( jo.has_array( "using" ) ) {
+        auto arr = jo.get_array( "using" );
+        r.reqs.clear();
+
+        while( arr.has_more() ) {
+            auto cur = arr.next_array();
+            r.reqs.emplace_back( requirement_id( cur.get_string( 0 ) ), cur.get_int( 1 ) );
+        }
     }
+
+    auto req_id = std::string( "inline_recipe_" ) += r.ident_;
+    requirement_data::load_requirement( jo, req_id );
+    r.reqs.emplace_back( requirement_id( req_id ), 1 );
 }
 
-void recipe_dictionary::clear()
+void recipe_dictionary::finalize()
 {
-    by_component.clear();
-    by_name.clear();
-    by_category.clear();
-    for( auto &recipe : recipes ) {
-        delete recipe;
+    for( auto it = recipe_dict.recipes.begin(); it != recipe_dict.recipes.end(); ) {
+        auto &r = it->second;
+        const char *id = it->first.c_str();
+
+        // concatenate requirements
+        r.requirements_ = std::accumulate( r.reqs.begin(), r.reqs.end(), requirement_data(),
+        []( const requirement_data & lhs, const std::pair<requirement_id, int> &rhs ) {
+            return lhs + ( *rhs.first * rhs.second );
+        } );
+
+        // remove blacklisted recipes
+        if( r.requirements().is_blacklisted() ) {
+            it = recipe_dict.recipes.erase( it );
+            continue;
+        }
+
+        // remove any invalid recipes...
+        if( !item::type_is_defined( r.result ) ) {
+            debugmsg( "Recipe %s defines invalid result", id );
+            it = recipe_dict.recipes.erase( it );
+            continue;
+        }
+
+        if( r.charges >= 0 && !item::count_by_charges( r.result ) ) {
+            debugmsg( "Recipe %s specified charges but result is not counted by charges", id );
+            it = recipe_dict.recipes.erase( it );
+            continue;
+        }
+
+        if( r.result_mult != 1 && !item::count_by_charges( r.result ) ) {
+            debugmsg( "Recipe %s has result_mult but result is not counted by charges", id );
+            it = recipe_dict.recipes.erase( it );
+            continue;
+        }
+
+        if( std::any_of( r.byproducts.begin(),
+        r.byproducts.end(), []( const std::pair<itype_id, int> &bp ) {
+        return !item::type_is_defined( bp.first );
+        } ) ) {
+            debugmsg( "Recipe %s defines invalid byproducts", id );
+            it = recipe_dict.recipes.erase( it );
+            continue;
```

Comment:
It would be nice to extract removal of invalid recipes into `std::remove_if()`. It would prevent this ugly repetition of 
```c++
it = recipe_dict.recipes.erase( it );
continue;
```
and the unpleasant `for( auto it = recipe_dict.recipes.begin(); it != recipe_dict.recipes.end(); ) {`. Also, that would be more structured.